### PR TITLE
git: fix darwin to linux cross-compilation

### PR DESCRIPTION
**Pull request by szlend**
Git uses `uname` for platform detection in some places which breaks cross-compilation from darwin to linux.

We can use the `deterministic-host-uname` package to replace `uname` with a mock mimicking the host's platform.

```
$ nix build -f . pkgsCross.aarch64-multiplatform.gitMinimal -L --print-out-paths
/nix/store/k25gxwp5xh06d70dxaw6ha33ajig5ssa-git-minimal-aarch64-unknown-linux-gnu-2.46.0
```

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/applications/version-management/git/default.nix [hunk 1]
```diff
@@ -84,7 +85,7 @@ stdenv.mkDerivation (finalAttrs: {
     done
   '';
 
-  nativeBuildInputs = [ gettext perlPackages.perl makeWrapper pkg-config ]
+  nativeBuildInputs = [ deterministic-host-uname gettext perlPackages.perl makeWrapper pkg-config ]
```
Comment (szlend) [inline pkgs/applications/version-management/git/default.nix:88]: I can also make this optional when `buildPlatform` and `hostPlatform` differ if that's preferred.

Comment (emilazy) [reply pkgs/applications/version-management/git/default.nix:88]: It’s better for native and cross builds to be the same whenever possible, so it should be fine as‐is.

**Comment by emilazy** [comment]
Just to check: We really want it for the *host* platform here, not the build platform? That’s kinda screwed up.

**Comment by szlend** [comment]
> Just to check: We really want it for the host platform here, not the build platform? That’s kinda screwed up.

@emilazy When compiling natively, `buildPlatform` == `hostPlatform`, so there shouldn't be any issues.

However when cross-compiling, you can see that the configuration scripts rely on `uname` a lot to make architectural build decisions:
- https://github.com/git/git/blob/ef8ce8f3d4344fd3af049c17eeba5cd20d98b69f/config.mak.uname
- https://github.com/git/git/blob/ef8ce8f3d4344fd3af049c17eeba5cd20d98b69f/config.mak.dev

In these cases we _want_ the configuration script to think that we're building for the `host` platform. From what I could gather there's no way to override this other than swapping in a mock `uname`.

**Comment by emilazy** [comment]
Yeah, okay. It might be better for us to patch those `uname_*` variables (or just set them in `makeFlags`? I forget if those take precedence over variables in the Makefile), and ideally upstream would fix this, but this minimal fix seems reasonable. This causes a lot of rebuilds, so please retarget to the `staging` branch.

**Comment by szlend** [comment]
Overriding these variables using `makeFlags` does work to my surprise. 

Something like this:
```
makeFlags = [
  "prefix=\${out}"
  "uname_S=${deterministic-host-uname.uSystem}"
  "uname_M=${deterministic-host-uname.processor}"
  "uname_O=${deterministic-host-uname.operatingSystem}"
  "uname_R=${deterministic-host-uname.modDirVersion}"
]
```

Though I feel a bit uncomfortable relying on internal details of the `deterministic-uname` derivation. I'm not sure if this is actually much better. Maybe `deterministic-uname` could expose a "public API" through `passthru`?

I could also compute these values myself, at the cost of adding a lot of complexity to an already large derivation.

File: pkgs/build-support/deterministic-uname/default.nix [hunk 2]
```diff
@@ -31,7 +31,7 @@ substituteAll {
     then "GNU/Linux"
     else if forPlatform.isDarwin
     then "Darwin" # darwin isn't in host-os.m4 so where does this come from?
-    else if stdenv.buildPlatform.isFreeBSD
+    else if forPlatform.isFreeBSD
```
Comment (szlend) [inline pkgs/build-support/deterministic-uname/default.nix:34]: This was also set incorrectly due to a bad merge.

File: pkgs/applications/version-management/git/default.nix [hunk 3]
```diff
@@ -130,8 +131,7 @@ stdenv.mkDerivation (finalAttrs: {
   # acceptable version.
   #
   # See https://github.com/Homebrew/homebrew-core/commit/dfa3ccf1e7d3901e371b5140b935839ba9d8b706
-  ++ lib.optional stdenv.hostPlatform.isDarwin "TKFRAMEWORK=/nonexistent"
-  ++ lib.optional (stdenv.hostPlatform.isFreeBSD && stdenv.hostPlatform != stdenv.buildPlatform) "uname_S=FreeBSD";
```
Comment (szlend) [inline pkgs/applications/version-management/git/default.nix:134]: No longer necessary.

**Review by philiptaron — APPROVED** [review]
Looks good!

**Comment by emilazy** [comment]
I’m fine with the current solution. If you could report it upstream to Git, that would be great.

**Comment by szlend** [comment]
I'm not sure if this actually needs an upstream issue. The platform _can_ be overwritten through `uname_*` make flags. Though it would sure be nice if this was documented at least.

**Comment by emilazy** [comment]
I guess the question is whether that’s intended as a public API or not. If it is, then I agree that we should consider simply having encodings of the relevant strings in `stdenv`, that could be used by the deterministic `uname` thing as well as by derivations like this. Anyway, going to merge this for now, but it would be great to follow up with the Git maintainers. Thanks!

**Pull request merged**